### ISSUE TITLE
Use detector perimeter information for cropping

### ISSUE THREAD

**Issue by joelvbernier** [HEXRD/hexrdgui#399]
In the shared Google Drive, I have uploaded detector perimeters for the TARDIS instrument (for now we will only deal with IMAGE-PLATE-[1,2]; IMAGE-PLATE-3 is cylindrically bent and I have to get that detector type finished).  The image data in the assoc. hdf5 archives is scanned into a 750 x 1250 array (100µm pixels), which is much bigger than the actual film.  Here is a view of ip4:
<img width="1680" alt="Screen Shot 2020-07-10 at 8 58 04 AM" src="https://user-images.githubusercontent.com/1154130/87174120-98526200-c28b-11ea-839a-015c28745862.png">
And here is a plot of the boundary, which is in physical coordinates (mm):
<img width="752" alt="Screen Shot 2020-07-10 at 9 01 17 AM" src="https://user-images.githubusercontent.com/1154130/87174431-01d27080-c28c-11ea-9037-7bed5e10db0e.png">
You can already see that a flip (mirror) about the horizontal axis is required in the image load.

What we need is the ability to translate and rotate (preferrably with mouse imput, but 3 sliders would work fine too) this contour on the raw image so the user can set up the the crop operation.  The actual cropping area with be the rectangle that circumscribes the image plate boundary.  The mask (valid values) would be the intersection of said rectangle and the area inside the image plate boundary.  Sound good?  Have fun :-)

**Comment by bnmajor** [comment]
This is great, thanks @joelvbernier!

**Comment by joelvbernier** [comment]
I'll get perimeters for the PXRDIP and BBXRD up as well -- those are the cases where all of the image plates (read: detectors) are scanned into a single big image, and the user will have to specify the crop for each detector's image sequentially.  The TARDIS case is easier in that there is one file per detector.

**Comment by bnmajor** [comment]
@joelvbernier Awesome, thank you. At this point I will proceed with the plan outlined in #384 to create a new load panel that will apply to this workflow, unless you see any problems with the current ideas.

**Comment by joelvbernier** [comment]
@bnmajor yeah, Let's call it an "Import Data" panel.  At this point, we could just have profiles for the different instruments (might be easiest way?).

**Comment by bnmajor** [comment]
@joelvbernier What will the config file look like for this use case? I'm curious what may be missing or need to be updated (like the rows and columns for each detector) after the cropping and detector masking has been applied.

**Comment by joelvbernier** [comment]
Good question!  The rows/cols have to match the crop box.  I imagine we can store a default instrument for each type, given that 1:1 correspondence.  I envision the user selecting which panel is active, then perfomring hte associated crop.


-Joel

On July 14, 2020 at 1:26 PM, bnmajor <notifications@github.com> wrote:






@joelvbernier What will the config file look like for this use case? I'm curious what may be missing or need to be updated (like the rows and columns for each detector) after the cropping and detector masking has been applied.

**Comment by bnmajor** [comment]
@joelvbernier Couple of questions related to this:

- I am assuming there will be cases where you will need to load multiple images at a time in order to create multi-frame image series. If this is the case will the template positioning for each detector be consistent across all frames, or will the detector outline need to be positioned for each frame?
- Will the raw `TARDIS` consistently be stored as contiguous arrays? Right now they cannot be read in with the `hdf5` adapter because it expects chunked arrays (this is being handled [here](https://github.com/HEXRD/hexrdgui/blob/57963389154e4fc1f3e9e0c59e6eb43115e131ce/hexrd/ui/image_file_manager.py#L81-L90) right now using `h5py` and creating the image series using the `array` adapter). If so we'll need to do additional manipulation to read in multiple files at once and I'm wondering if this should continue to be handled in `hexrdgui` or if `hexrd` should be extended to handle this case? Or if I've overlooked an option that already exists?
- We still need detector perimeters for `PXRDIP` and `BBXRD` to test cropping and masking for those instruments. If it is possible to get config files for them as well that would be helpful too.

**Comment by joelvbernier** [comment]
@bnmajor 

for the LLNL workflow, these are all ”single shot”, so there is only one image per panel.  There might be e exception with a new instrument in which the detectors are multiplexed to grab a burst of frames (~2-10), but I don't have an example yet.  Luckily, those detectors usually aren't irregular in shape like the image plates￼.  I’d say the toughest cases in the single-shot paradigm are TARDIS, PXRDIP, and BBXRD.￼

**Comment by bnmajor** [comment]
@joelvbernier In order to use the same load path for all data we do need to use an imageseries - this could certainly be worked around but the current fix seems to work fine and be relatively simple so unless it becomes a more complicated problem or we decide to catch the case in the adapter I will just leave it as it is for now.

Thanks for the help!

**Comment by joelvbernier** [comment]
I think we'll make the edit to the imageseries module anyhow, but I'll consult with @donald-e-boyce and keep you posted.

Regarding the TARDIS, as I just spend the last 3 days analyzing data...

Attached are updated boundary files to represent the proper origin and orientation of the local detector cartesian coordinates (plus some useful header information).  The bounding boxes for the 3 images are as follows (for the standard 100µm scan pitch):

IMAGE-PLATE-2: (421, 901)
IMAGE-PLATE-3: (234, 523) **don't use this one in the instrument yet
IMAGE-PLATE-2: (421, 901)

Here my current workflow:

1. load the raw HDF5 image data from the NIF archive for each image plate
2. flip them all about the horizontal axis
3. IMAGE-PLATE-2 _usually_ is scanned in with the same rough orientation as the crop boundary (semicircle facing up), so I just crop 'n go
4. Same goes for IMAGE-PLATE-3...
5. IMAGE-PLATE-4 is usually scanned in rotated 180° from the reference crop boundary (the new one I attached), so I flip 'h', then rotate 180 before cropping.

I don't currently have the ability to rotate the crop, so I have to take care of small misalignment in the calibration.

So it looks like those would be the default settings for importing TARDIS data.  I have also included the default TARDIS instrument (no IMAGE-PLATE-3 yet...).

lmk if that makes sense.

[TARDIS_IMAGE-PLATE-3_bnd.txt](https://github.com/HEXRD/hexrdgui/files/5008060/TARDIS_IMAGE-PLATE-3_bnd.txt)
[TARDIS_IMAGE-PLATE-4_bnd.txt](https://github.com/HEXRD/hexrdgui/files/5008062/TARDIS_IMAGE-PLATE-4_bnd.txt)
[TARDIS_IMAGE-PLATE-2_bnd.txt](https://github.com/HEXRD/hexrdgui/files/5008063/TARDIS_IMAGE-PLATE-2_bnd.txt)

```
beam:
  energy: 10.2505
  vector:
    azimuth: 60.6212291803227
    polar_angle: 90.21969555888647
detectors:
  IMAGE-PLATE-2:
    pixels:
      columns: 901
      rows: 421
      size:
      - 0.1
      - 0.1
    buffer:
    - 0.2
    - 0.2
    transform:
      tilt:
      - 0.0
      - -2.22144147e+00
      - 2.22144147e+00
      translation:
      - 0.0
      - -11.935
      - -25.08
  IMAGE-PLATE-4:
    pixels:
      columns: 901
      rows: 421
      size:
      - 0.1
      - 0.1
    buffer:
    - 0.2
    - 0.2
    transform:
      tilt:
      - -1.57079633e+00
      - 0.0
      - 0.0
      translation:
      - 0.0
      - 11.935
      - -25.08
id: NXXXXXX-XXX
oscillation_stage:
  chi: 0.0
  translation:
  - 0.0
  - 0.0
  - 0.0
```

**Comment by bnmajor** [comment]
While we are continuing to improve on the import data panel, this issue was resolved by #410. Closing.
